native/membrane_videocompositor/src/elixir_bridge/elixir_structs.rs

```diff
@@ -136,3 +140,710 @@ impl From<ElixirCropping> for Box<dyn TextureTransformation> {
+        assert!(result.is_err());
+        if let Err(SceneParsingError::UnusedObject(name)) = result {
+            assert_eq!(name, d);
+        } else {
+            panic!("Parser failed to detect unused objects")
+        }
+    }
+
+    #[test]
+    fn scene_parsing_simple_success() {
+        let a = Name::Atom("a".into());
+        let b = Name::Atom("b".into());
+        let c = Name::Atom("c".into());
+
+        let inputs = HashMap::from_iter(vec![(a.clone(), a.clone()), (b.clone(), b.clone())]);
+
+        let scene = Scene {
+            objects: vec![
+                (
+                    a.clone(),
+                    Object::Video(InputVideo {
+                        input_pad: "pad1".into(),
+                    }),
+                ),
+                (
+                    b,
+                    Object::Video(InputVideo {
+                        input_pad: "pad2".into(),
+                    }),
+                ),
+                (
+                    c.clone(),
+                    Object::Layout(Layout {
+                        inputs,
+                        resolution: LayoutOutputResolution::Name(a),
+                        implementation: 42,
+                    }),
+                ),
+            ],
+            output: c,
+        };
+
+        let result: Result<crate::scene::Scene, _> = scene.try_into();
+        assert!(result.is_ok())
+    }
+
+    #[test]
+    fn scene_parsing_maintains_transformation_order() {
+        let a = Name::Atom("a".into());
+        let b = Name::Atom("b".into());
+
+        let scene = Scene {
+            objects: vec![
+                (
+                    a.clone(),
+                    Object::Video(InputVideo {
+                        input_pad: "pad1".into(),
+                    }),
+                ),
+                (
+                    b.clone(),
+                    Object::Texture(Texture {
+                        input: a,
+                        resolution: TextureOutputResolution::TransformedInputResolution,
+                        transformations: vec![0, 1, 2],
+                    }),
+                ),
+            ],
+            output: b,
+        };
+
+        let result: Result<crate::scene::Scene, _> = scene.try_into();
+        let scene = result.unwrap();
+
+        use crate::scene::Node;
+
+        let final_node = scene.final_node;
+
+        let Node::Transformation { ref previous, transformation, .. } = *final_node else {
+            panic!("unexpected scene structure");
+        };
+        assert_eq!(transformation, 2);
+
+        let Node::Transformation { ref previous, transformation, .. } = **previous else {
+            panic!("unexpected scene structure");
+        };
+        assert_eq!(transformation, 1);
+
+        let Node::Transformation { transformation, .. } = **previous else {
+            panic!("unexpected scene structure");
+        };
+        assert_eq!(transformation, 0);
+    }
+}
```

Comment:
This module does literary everything. Maps elixir structs, handle rustler weird stuff, validate scene, defines validation and parsing errors ... Basically, responsibilities separation doesn't exist here at all. It's just a `one module to rule them all`. I would separate it, such that:
- dealing with rustler stuff is handled separately
- validation is separated from translation 

![download (2)](https://user-images.githubusercontent.com/104033489/231432818-6fc2553a-d5e5-4c6f-8470-6eaaaf4f5e1e.png)

Reply:
i split it

Reply:
Well, it's still weird to have WIP `Scene` modules in the same module with working stuff like `ElixirBaseVideoPlacement`, specific to the current VC version.

Reply:
perhaps the work on the new compositor should be moved away from master into a different branch